lib/ex_doc/formatter/epub.ex

```diff
@@ -0,0 +1,231 @@
+    generate_toc(output, config, modules, exceptions, protocols, uuid)
+    generate_nav(output, config, modules, exceptions, protocols)
+    generate_title(output, config)
+    generate_list(output, config, modules)
+    generate_list(output, config, exceptions)
+    generate_list(output, config, protocols)
+
+    {:ok, epub_file} = generate_epub(output, config)
+    delete_extras(output)
+
+    epub_file
+  end
+
+  defp generate_mimetype(output) do
+    content = "application/epub+zip"
+    File.write("#{output}/mimetype", content)
+  end
+
+  defp generate_extras(output, config, module_nodes) do
+    config.extras
+    |> Enum.map(&Task.async(fn ->
+         generate_extra(&1, output, config, module_nodes)
+       end))
+    |> Enum.map(&Task.await/1)
+  end
+
+  defp generate_extra(input, output, config, module_nodes) do
+    file_ext =
+      input
+      |> Path.extname
+      |> String.downcase
+
+    if file_ext in [".md"] do
+      file_name =
+        input
+        |> Path.basename(".md")
+        |> String.upcase
+
+      content =
+        input
+        |> File.read!
+        |> HTML.Autolink.project_doc(module_nodes)
+
+      config = Map.put(config, :title, file_name)
+      extra_html =
+        Templates.extra_template(config, content)
+        |> valid_xhtml_ids
+
+      File.write!("#{output}/OEBPS/modules/#{file_name}.html", extra_html)
+    else
+      raise ArgumentError, "file format not recognized, allowed format is: .md"
+    end
+  end
+
+  defp generate_content(output, config, modules, exceptions, protocols, uuid, datetime) do
+    content = Templates.content_template(config, modules ++ exceptions ++ protocols, uuid, datetime)
+    File.write("#{output}/OEBPS/content.opf", content)
+  end
+
+  defp generate_toc(output, config, modules, exceptions, protocols, uuid) do
+    content = Templates.toc_template(config, modules ++ exceptions ++ protocols, uuid)
+    File.write("#{output}/OEBPS/toc.ncx", content)
+  end
+
+  defp generate_nav(output, config, modules, exceptions, protocols) do
+    content = Templates.nav_template(config, modules ++ exceptions ++ protocols)
+    File.write("#{output}/OEBPS/nav.html", content)
+  end
+
+  defp generate_title(output, config) do
+    content = Templates.title_template(config)
+    File.write("#{output}/OEBPS/title.html", content)
+  end
+
+  defp generate_list(output, config, nodes) do
+    nodes
+    |> Enum.map(&Task.async(fn -> generate_module_page(output, config, &1) end))
+    |> Enum.map(&Task.await/1)
+  end
+
+  defp generate_epub(output, config) do
+    output = Path.expand(output)
+    target_path =
+      "#{output}/#{config.project}-v#{config.version}.epub"
+      |> String.to_char_list
+
+    {:ok, zip_path} = :zip.create(target_path,
+                                  files_to_add(output),
+                                  compress: ['.css', '.html', '.ncx', '.opf',
+                                             '.jpg', '.png', '.xml'])
+    {:ok, zip_path}
```

Comment:
That's a common Elixir practice. Somtimes it is preferable to match on the result when we call it so we get the proper cause on errors: http://blog.plataformatec.com.br/2014/09/writing-assertive-code-with-elixir/